### ISSUE TITLE
Proposed no-magic-numbers option: ignoreJSXNumbers

### ISSUE THREAD

**Issue by bsbeeks** [eslint/eslint#5348]
eslint version: 2.1.0

When using the `no-magic-numbers` rule, numbers that are placed inside of JSX props are caught by the rule. Example:
```javascript
<input maxLength={10} />
```
returns
```javascript
error  No magic number: 10   no-magic-numbers
```
Assigning numbers to props is similar to assigning them to variables, where the name of the prop makes the meaning explicit. In these cases the rule should be ignored. I propose a new option is added to this rule called `ignoreJSXNumbers`. This option would ignore any number that is defined as a JSX element property.

**Comment by platinumazure** [comment]
I think this has been discussed before. If you just put the number in quotes (and not in braces), then it won't be parsed as a numeric literal. I don't know JSX though, so maybe there's a reason that wouldn't work.

**Comment by bsbeeks** [comment]
Typically with JSX and React you define `propTypes` to validate component props. One of these is `React.PropTypes.number` which requires the prop to be a number and not a string. In those cases (which are common) using quotes won't work.

**Comment by ilyavolodin** [comment]
Not familiar enough with React to comment. @btmills @xjamundx could you comment please? It feels like a valid case, as I know `propTypes` are considered one of the best practices for React. On the other hand, ESLint doesn't support React, and this is not part of the JSX spec, so maybe it should be a request for `eslint-plugin-react` instead of core?

**Comment by bsbeeks** [comment]
@ilyavolodin that's a good point. I was thinking since ESLint supports JSX it would live here, but given the React `propTypes` scenario adding it to `eslint-plugin-react` would make more sense. If that's the case, is there a way to extend/add to this rule from within the plugin? It seems overkill to create an entirely new rule for something that can be solved with a rule option.

**Comment by ilyavolodin** [comment]
@bsbeeks Unfortunately we don't currently have a mechanism to extend existing rules. It's been brought up a few times before, but we don't have any concrete proposals for solving this now. So in `eslint-plugin-react` it would have to be a new rule.

**Comment by xjamundx** [comment]
My 2 cents is that we should exclude JSX from these checks, by default.

Do we ever want magic number checking in JSX?

Why should it warn on  `<input maxLength={10} />`?

**Comment by ilyavolodin** [comment]
@xjamundx Because 10 is a magic number?:-)

**Comment by bsbeeks** [comment]
@xjamundx I agree. The only use case I can think of where I'd want it to check is if the number is inside an array `<MyComponent someProp={[2, 6]} />`. In that case I'd want those numbers assigned to variables.

**Comment by xjamundx** [comment]
haha I guess I don't know what a "magic number" is for this rule.

**Comment by nzakas** [comment]
Hour about if we omit JSX expressions that only contain a single number, and nothing else, by default? Otherwise, the rule still applies.

**Comment by bsbeeks** [comment]
@nzakas I can roll with that. Let me know if that's the direction we want to head and I'll update my PR.

**Comment by nzakas** [comment]
I think that's a good direction, though I'd like to hear from @xjamundx to confirm.

**Comment by ilyavolodin** [comment]
@xjamundx Any insights?

**Comment by xjamundx** [comment]
Sounds reasonable 👍🏼

**Comment by nzakas** [comment]
Okie. @bsbeeks if you'd like to submit a PR, this is ready

**Comment by bsbeeks** [comment]
@nzakas sounds good. I should have some time tonight to modify https://github.com/eslint/eslint/pull/5347

**Comment by bsbeeks** [comment]
@nzakas @xjamundx Updated #5347 to reflect the changes discussed. Let me know if any additional adjustments need to be made.
